akka-stream/src/main/scala/akka/stream/impl/io/InputStreamSinkStage.scala

```diff
@@ -166,18 +169,17 @@ private[akka] class InputStreamAdapter(sharedBuffer: BlockingQueue[StreamToAdapt
                             gotBytes: Int): Int = {
     grabDataChunk() match {
       case Some(data) ⇒
-        val size = data.size - skipBytes
-        if (size + gotBytes <= length) {
-          System.arraycopy(data.toArray, skipBytes, arr, begin, size)
-          skipBytes = 0
+        val size = data.size
+        if (size <= length) {
+          data.copyToArray(arr, begin, size)
           detachedChunk = None
-          if (length - size == 0)
+          if (size == length)
             gotBytes + size
           else
             getData(arr, begin + size, length - size, gotBytes + size)
         } else {
-          System.arraycopy(data.toArray, skipBytes, arr, begin, length)
-          skipBytes = length
+          data.copyToArray(arr, begin, length)
+          detachedChunk = Some(data.drop(length))
```

Comment:
`detachedChunk` does not need to be an Option[ByteString], it can be a byteString where the empty ByteString is == None.